Add optional type hinting to arguments

DominicOram opened the pull request:
# Description

Adds `Optional` type hint in all places where arguments default to `None` and so are clearly `Optional`

Fixes https://github.com/open-telemetry/opentelemetry-python/issues/4454

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

1. Temporarily comment out `typeCheckingMode = "off"` in the `pyproject.toml`
2. Run `pyright .` on `main` and see 1560 errors
3. Run `pyright .` on this branch and see 1508 errors

# Does This PR Require a Contrib Repo Change?

Don't think so?

- [ ] Yes. - Link to PR: 
- [ ] No.

# Checklist:

- [ ] Followed the style guidelines of this project
- [ ] Changelogs have been updated
- [ ] Unit tests have been added
- [ ] Documentation has been updated

xrmx reviewed (commented):
If you add `from __future__ import annotations` you can avoid importing Optional and use `| None` syntax, while you are at it you may also drop the usage of `Dict` and `List` and use `dict` and `list` instead.

DominicOram commented:
> If you add `from __future__ import annotations` you can avoid importing Optional and use `| None` syntax, while you are at it you may also drop the usage of `Dict` and `List` and use `dict` and `list` instead.

Updated the places where I have changed something and a few places in the same area to keep some consistency. I would be happy to do more but feels like a new PR as it would be a big change and I would rather get my original issue fixed sooner

xrmx commented on shim/opentelemetry-opentracing-shim/src/opentelemetry/shim/opentracing_shim/__init__.py:
```diff
@@ -613,11 +614,11 @@ def start_active_span(
-        operation_name: str = None,
-        child_of: Union[SpanShim, SpanContextShim] = None,
-        references: list = None,
+        operation_name: Optional[str] = None,
```
missed this hunk

xrmx commented on opentelemetry-sdk/src/opentelemetry/sdk/_logs/_internal/export/__init__.py:
```diff
@@ -171,10 +171,10 @@ class BatchLogRecordProcessor(LogRecordProcessor):
-        max_export_batch_size: int = None,
-        export_timeout_millis: float = None,
-        max_queue_size: int = None,
+        schedule_delay_millis: Optional[float] = None,
```
missed this hunk

xrmx commented on shim/opentelemetry-opentracing-shim/src/opentelemetry/shim/opentracing_shim/__init__.py:
```diff
@@ -613,11 +614,11 @@ def start_active_span(
-        operation_name: str = None,
-        child_of: Union[SpanShim, SpanContextShim] = None,
-        references: list = None,
+        operation_name: Optional[str] = None,
+        child_of: Optional[Union[SpanShim, SpanContextShim]] = None,
+        references: Optional[list] = None,
```
```suggestion
        operation_name: str | None = None,
        child_of: SpanShim | SpanContextShim | None = None,
        references: list | None = None,
```

xrmx commented on shim/opentelemetry-opentracing-shim/src/opentelemetry/shim/opentracing_shim/__init__.py:
```diff
@@ -613,11 +614,11 @@ def start_active_span(
+        references: Optional[list] = None,
         tags: Attributes = None,
-        start_time: float = None,
+        start_time: Optional[float] = None,
```
```suggestion
        start_time: float | None = None,
```

xrmx commented:
I've filed this https://github.com/open-telemetry/opentelemetry-python-contrib/issues/3320 because this PR failed at least a couple of times that test. Since this continues to fail maybe it's not a flaky test and we are introducing a regression somewhere here instead.
